music21/osmd/osmd.py

```diff
@@ -0,0 +1,233 @@
+# -*- coding: utf-8 -*-
+# -------------------------------------------------------------------------------
+# Name:         osmd/osmd.py
+# Purpose:      display Streams in IPython notebooks using Open Sheet Music Display
+#
+# Authors:      Sven Hollowell
+#
+# Copyright:    Copyright © 2018 Michael Scott Cuthbert and the music21 Project
+# License:      LGPL or BSD, see license.txt
+# -------------------------------------------------------------------------------
+'''
+Registers the .show('osmd') converter for use in IPython notebook
+If not in a notebook (e.g. commandline), it will open a browser window with the rendered score
+See OSMD project page here: https://github.com/opensheetmusicdisplay/opensheetmusicdisplay
+'''
+import unittest
+import time
+import random
+import json
+import os
+import importlib
+import tempfile
+import urllib.request
+import webbrowser
+
+from music21.converter.subConverters import SubConverter
+from music21.instrument import Piano
+from music21.common import getSourceFilePath, runningUnderIPython
+from music21 import exceptions21
+from music21.musicxml import m21ToXml
+from music21 import environment
+
+# when using offline=True, sets the disk location where we will cache the OSMD.min.js file:
+environLocal = environment.Environment('osmd.osmd')
+
+class OpenSheetMusicDisplayException(exceptions21.Music21Exception):
+    pass
+
+
+# Determine if we are running inside an IPython notebook
+# if so will render inline - if not we open a new browser window
+try:
+    loader = importlib.util.find_spec('IPython.core.display')
+except ImportError:
+    loader = None
+hasInstalledIPython = loader is not None
+
+
+def getExtendedModules():
+    # pylint: disable=import-outside-toplevel
+    if hasInstalledIPython:
+        from IPython.core.display import display, HTML, Javascript
+    else:
+        def display():
+            raise OpenSheetMusicDisplayException(
+                'OpenSheetMusicDisplay requires IPython to be installed')
+
+        HTML = Javascript = display
+    return display, HTML, Javascript
+
+
+class ConverterOpenSheetMusicDisplay(SubConverter):
+    '''
+    Converter for displaying sheet music inline
+    in an IPython notebook.
+    Uses: https://opensheetmusicdisplay.org/ for rendering
+    '''
+    registerFormats = ('osmd',)
+    registerShowFormats = ('osmd',)
+    # when updating the script tag osmd will only reload in a notebook if you: Kernel -> restart and
```

Comment:
```suggestion
    registerOutputFormats = ('html',)
```
Then you can call `self.getTemporaryFile()`. It was staring at me when I compiled the osmd module documentation.

Reply:
Obviously don't get rid of the comment, I guess you can't suggest insertions?

Reply:
Ah I see how it works now, had to change the line to `registerOutputExtensions = ('html',)` and then it worked :)

Reply:
Yes this is working now, seems like default on windows is `%HOME%/AppData/Local/Temp/music21`

Reply:
Yeah, there's some tooling for this in music21, you can set the music21 environment variable `directoryScratch` to put it somewhere else. For users who have done so, I'm trying to avoid "where did it go?" posts.